DepotDownloader/TotpAuthenticator.cs

```diff
@@ -0,0 +1,91 @@
+using System;
+using System.Security.Cryptography;
+using System.Threading.Tasks;
+using SteamKit2.Authentication;
+
+namespace DepotDownloader;
+
+/// <summary>
+/// Implementation of <see cref="IAuthenticator"/> that uses a TOTP Authenticator key to generate TOTP verification codes.
+/// Falls back to the provided fallback authenticator.
+/// </summary>
+public class TotpAuthenticator : IAuthenticator
+{
+    private readonly string _totpKey;
+    private readonly IAuthenticator _fallbackAuthenticator;
+
+    public TotpAuthenticator(string totpKey, IAuthenticator fallbackAuthenticator)
+    {
+        _totpKey = totpKey;
+        _fallbackAuthenticator = fallbackAuthenticator;
+    }
+
+    /// <inheritdoc />
+    public Task<string> GetDeviceCodeAsync(bool previousCodeWasIncorrect)
+    {
+        if (previousCodeWasIncorrect)
+        {
+            return _fallbackAuthenticator.GetDeviceCodeAsync(true);
+        }
+
+        var deviceCode = GetDeviceCode(_totpKey);
+        return deviceCode != null ? Task.FromResult(deviceCode) : _fallbackAuthenticator.GetDeviceCodeAsync(false);
+    }
+
+    /// <inheritdoc />
+    public Task<string> GetEmailCodeAsync(string email, bool previousCodeWasIncorrect)
+    {
+        return _fallbackAuthenticator.GetEmailCodeAsync(email, previousCodeWasIncorrect);
+    }
+
+    /// <inheritdoc />
+    public Task<bool> AcceptDeviceConfirmationAsync()
+    {
+        return _fallbackAuthenticator.AcceptDeviceConfirmationAsync();
+    }
+
+    // https://github.com/bitwarden/mobile/blob/7a65bf7fd7b44424073201c2c574d45b64b9ec9d/src/Core/Services/TotpService.cs
```

Comment:
Bitwarden is GPLv3. I don't believe we can use this code in DD, which is a GPLv2 project.

Reply:
Damn I overlooked that. I can probably swap it out with this which is MIT licensed.
https://github.com/geel9/SteamAuth/blob/4ccfa2e9258dda67f6e8baefe3e2833419a565de/SteamAuth/SteamGuardAccount.cs#L88